### PR TITLE
[RFC] [partition] FS choice for autopartition options

### PR DESCRIPTION
**Pull request by LordTermor**
Allows to choose filesystem when _erase disk_ is chosen (and potentially other autopartition options) and configure available filesystems using `partition.conf`.
изображение
изображение

### CODE REVIEW

File: src/modules/partition/partition.conf [hunk 1]
```diff
@@ -137,7 +137,7 @@ initialSwapChoice: none
 # If nothing is specified, Calamares defaults to "ext4".
 #
 # Names are case-sensitive and defined by KPMCore.
-defaultFileSystemType:  "ext4"
+availableFileSystemTypes:  ["ext4", "btrfs", "f2fs"]
```
Comment (Conan-Kudo) [inline src/modules/partition/partition.conf:140]: Does this imply order/preference?

**Review by Conan-Kudo — CHANGES_REQUESTED** [review]

**Comment by Conan-Kudo** [comment]
Also, if we're going to have _choices_ for autopart options, how do we define schemes for each of them?

File: src/modules/partition/partition.conf (hunk 1 above)
Comment (abucodonosor) [reply src/modules/partition/partition.conf:140]: I don't think so but this should not be like this.

defaultFileSystemType should remain as is, IOW this is what the installer defaults too,
and availableFileSystemType should be called something like optionalFileSystemTypes.

**Comment by abucodonosor** [comment]
> Also, if we're going to have _choices_ for autopart options, how do we define schemes for each of them?

Not possible with the current implementation. So this option should conflict with the schemes one, until
the scheme one gets UIX bits I guess.

File: src/modules/partition/partition.conf (hunk 1 above)
Comment (Conan-Kudo) [reply src/modules/partition/partition.conf:140]: That makes more sense to me, yes.

Comment (LordTermor) [reply src/modules/partition/partition.conf:140]: Well it's still a draft and I wanted to discuss the name with the rest of contributors. I don't understand why there should be two config values? We can leave only one and use it as a string in cases when there should be only one option.

**Comment by LordTermor** [comment]
> 
> 
> > Also, if we're going to have _choices_ for autopart options, how do we define schemes for each of them?
> 
> Not possible with the current implementation. So this option should conflict with the schemes one, until
> the scheme one gets UIX bits I guess.

Where can I read more about schemes?

**Comment by abucodonosor** [comment]
> > > Also, if we're going to have _choices_ for autopart options, how do we define schemes for each of them?
> > 
> > 
> > Not possible with the current implementation. So this option should conflict with the schemes one, until
> > the scheme one gets UIX bits I guess.
> 
> Where can I read more about schemes?

See the partition.conf & partitionLayout section.

File: src/modules/partition/partition.conf (hunk 1 above)
Comment (abucodonosor) [reply src/modules/partition/partition.conf:140]: If you want to remove the default option then you have to change a lot of logic all over the place,
including the forced usage of ext4 in the case of missing default FS. 
Not to mention by just removing it without deprecation or automagically replace,
all existing downstream calamares configuration packages will break, or do at least very unexpected things.

Just do something like 

git grep defaultFileSystemType
git grep ext4

in the calamares source code.

Comment (hphilm) [reply src/modules/partition/partition.conf:140]: > I don't think so but this should not be like this.
> 
> defaultFileSystemType should remain as is, IOW this is what the installer defaults too,
> and availableFileSystemType should be called something like optionalFileSystemTypes.

Isn't that not the exact request you did [2017](https://github.com/calamares/calamares/issues/847) regarding that topic? Default is the first FS you specify. If none is specified it defaults to Ext4.

**Comment by hphilm** [comment]
Seems we might need adjustments to the [template schema](https://github.com/calamares/calamares/blob/calamares/src/modules/partition/partition.schema.yaml#L20) anyway, regarding default FS type ...

File: src/modules/partition/partition.conf (hunk 1 above)
Comment (LordTermor) [reply src/modules/partition/partition.conf:140]: > If you want to remove the default option then you have to change a lot of logic all over the place,
> including the forced usage of ext4 in the case of missing default FS.
> Not to mention by just removing it without deprecation or automagically replace,
> all existing downstream calamares configuration packages will break, or do at least very unexpected things.
> 
> Just do something like
> 
> git grep defaultFileSystemType
> git grep ext4
> 
> in the calamares source code.

Well deprecation thing is ok. It's not too much places where defaultFileSystemType is used I think I can address them all. I just think that leaving and using two options is a bit dirty.

Comment (abucodonosor) [reply src/modules/partition/partition.conf:140]: At the time nothing like the partition scheme existed among many other changes in this area.
But what I requested there is an additional list of supported FSes by the distribution.

Comment (abucodonosor) [reply src/modules/partition/partition.conf:140]: > Well deprecation thing is ok. It's not too much places where defaultFileSystemType is used I think I can address them all. I just think that leaving and using two options is a bit dirty.

Or maybe hide it in the code.. Eg: we remove the configurationMap check for it, then set it to the first fs in the new_option_of_FSes, with a fallback to ext4 if everything fails ( config bad or whatever else ), that way not too 
much logic needs to change.

File: src/modules/partition/core/PartitionActions.cpp [hunk 2]
```diff
@@ -109,6 +109,10 @@ doAutopartition( PartitionCoreModule* core, Device* dev, Choices::AutoPartitionO
         partType = isEfi ? PartitionTable::gpt : PartitionTable::msdos;
     }
 
+    FileSystem::Type type;
+    PartUtils::findFS(o.defaultFsType, &type);
+    core->initLayout( type == FileSystem::Unknown ? FileSystem::Ext4 : type);
```
Comment (hphilm) [inline src/modules/partition/core/PartitionActions.cpp:114]: I think the fallback filesystem is addressed here ...

File: src/modules/partition/partition.conf (hunk 1 above)
Comment (adriaandegroot) [reply src/modules/partition/partition.conf:140]: The point is that **existing** configs which set *defaultFileSystem* now don't set anything at all. The configuration (`setConfigurationMap()`) code could check for that key and decide it means a one-element list, but that does need to be spelled out.

File: src/modules/partition/core/PartitionActions.cpp (hunk 2 above)
Comment (LordTermor) [reply src/modules/partition/core/PartitionActions.cpp:114]: It's for the case when fs string in config is invalid

File: src/modules/partition/core/Config.cpp [hunk 3]
```diff
@@ -207,6 +207,15 @@ Config::setSwapChoice( Config::SwapChoice c )
     }
 }
 
+void
+Config::setEraseFsTypeChoice(const QString& choice)
```
Comment (adriaandegroot) [inline src/modules/partition/core/Config.cpp:211]: Apply the coding-style tool (I'll just say that once to get it over with; `ci/calamaresstyle <dirname>` will do the trick, assuming you have astyle + clang-format10 or so.

File: src/modules/partition/core/Config.cpp [hunk 4]
```diff
@@ -207,6 +207,15 @@ Config::setSwapChoice( Config::SwapChoice c )
...
+{
+    if(choice != m_eraseFsTypeChoice){
+        m_eraseFsTypeChoice = choice;
```
Comment (adriaandegroot) [inline src/modules/partition/core/Config.cpp:214]: I realise that this "config" object does not follow the style of most other config classes, but normally there would be an `emit eraseFsTypeChoiceChanged()` here. It *might* make sense to start introducing that style to the config object.

File: src/modules/partition/core/Config.cpp [hunk 5]
```diff
@@ -207,6 +207,15 @@ Config::setSwapChoice( Config::SwapChoice c )
     }
 }
 
+void
+Config::setEraseFsTypeChoice(const QString& choice)
+{
+    if(choice != m_eraseFsTypeChoice){
```
Comment (adriaandegroot) [inline src/modules/partition/core/Config.cpp:213]: Is there an idea of "valid choices"? Will any string do?

File: src/modules/partition/core/Config.cpp [hunk 6]
```diff
@@ -240,6 +249,17 @@ Config::setConfigurationMap( const QVariantMap& configurationMap )
     gs->insert( "allowManualPartitioning",
                 CalamaresUtils::getBool( configurationMap, "allowManualPartitioning", true ) );
 
+    if ( configurationMap.contains( "availableFileSystemTypes" ) ) {
+        if(configurationMap.value(  "availableFileSystemTypes" ).type() == QVariant::List ){
+            m_eraseFsTypes.clear();
+            m_eraseFsTypes.append(configurationMap.value(  "availableFileSystemTypes" ).toStringList());
```
Comment (adriaandegroot) [inline src/modules/partition/core/Config.cpp:255]: These two lines can be one assignment instead of clear + append; you could use `getStringList()` although it is a bit wordy.

File: src/modules/partition/core/Config.cpp [hunk 7]
```diff
@@ -240,6 +249,17 @@ Config::setConfigurationMap( const QVariantMap& configurationMap )
...
+            m_eraseFsTypeChoice = m_eraseFsTypes[0];
```
Comment (adriaandegroot) [inline src/modules/partition/core/Config.cpp:256]: You can lift this one out of (both branches of the) if and put it after the is-this-a-list-or-a-single-string checks.

Comment (adriaandegroot) [inline src/modules/partition/core/Config.cpp:256]: If the list is empty this is still UB (e.g. if the user has the terrible configuration `availableFileSystemTypes: []`)

File: src/modules/partition/core/Config.cpp [hunk 8]
```diff
@@ -240,6 +249,17 @@ Config::setConfigurationMap( const QVariantMap& configurationMap )
...
+        } else if ( configurationMap.value(  "availableFileSystemTypes" ).type() == QVariant::String  ){
+            m_eraseFsTypes.append(configurationMap.value(  "availableFileSystemTypes" ).toString());
```
Comment (adriaandegroot) [inline src/modules/partition/core/Config.cpp:258]: Why **isn't** the list cleared here, first? It could be an assignment, too.

File: src/modules/partition/core/Config.cpp [hunk 9]
```diff
@@ -240,6 +249,17 @@ Config::setConfigurationMap( const QVariantMap& configurationMap )
...
+            m_eraseFsTypeChoice = m_eraseFsTypes[0];
```
Comment (adriaandegroot) [inline src/modules/partition/core/Config.cpp:259]: .. come to think of it, calling the setter is probably better, since you might want to emit a signal; this depends a little bit on what's in the setter and whether everything is fully set-up for calling it yet.

File: src/modules/partition/gui/ChoicePage.h [hunk 10]
```diff
@@ -139,6 +139,8 @@ private slots:
     Calamares::PrettyRadioButton* m_replaceButton;
     Calamares::PrettyRadioButton* m_somethingElseButton;
     QComboBox* m_eraseSwapChoiceComboBox;  // UI, see also m_eraseSwapChoice
+    QComboBox* m_eraseFsTypesChoiceComboBox;  // UI, see also m_eraseSwapChoice
```
Comment (adriaandegroot) [inline src/modules/partition/gui/ChoicePage.h:142]: Comment mismatch. Also, start to prefer `= nullptr` initializations for pointers (esp. relevant because you didn't initialize it in the constuctor).

File: src/modules/partition/gui/ChoicePage.cpp [hunk 11]
```diff
@@ -246,6 +246,14 @@ ChoicePage::setupChoices()
         m_eraseButton->addOptionsComboBox( m_eraseSwapChoiceComboBox );
     }
 
+    if ( m_config->eraseFsTypes().count() > 1)
+    {
+        m_eraseFsTypesChoiceComboBox = new QComboBox;
+        m_eraseFsTypesChoiceComboBox->addItems(m_config->eraseFsTypes());
+        connect(m_eraseFsTypesChoiceComboBox, &QComboBox::currentTextChanged, [this](const auto& currentString){m_config->setEraseFsTypeChoice(currentString); onActionChanged();});
```
Comment (adriaandegroot) [inline src/modules/partition/gui/ChoicePage.cpp:253]: This looks like a case for:
- connecting directly to `&Config::setEraseFsTypeChoice`
- giving the config a signal to .. signal changes to the erase choice
- connecting the change signal to `onActivate()`

**Comment by adriaandegroot** [comment]
Open conversation topic: **existing configurations**

- An existing configuration can (should) have a *defaultFileSystemType*.
- If that is set and no erase-fs-type is set, then what? (I'd interpret it as "one item erase-fs-type list")
- if it is set and also erase-fs-type is set, then what? (I'd interpret it as "use erase-fs-type list and set the selected one to the configured default, rather than the first in the list")

To land this before 3.3 (which is scheduled "sometime, eventually") you'll need to address backwards-compatibility like that. For 3.3 it's possible to "just" redefine the config meanings.

**Comment by adriaandegroot** [comment]
Open conversation topic: **documentation**

The config file should be extensively documented; there should be examples. If the order of a list matters, there should be comments explaining that. There is a `partition.schema.yaml` which is the schema file for the YAML configuration. This needs to be updated to match the new config keys.

**Comment by adriaandegroot** [comment]
Open conversation topic: **partition schema**

The key *partitionLayout* in the config file defines .. a configuration to apply when erasing. Each entry has a *name* and a *filesystem*. The interation with *defaultFileSystemType* is already a bit odd in **that** scheme, since it completely ignores the default FS type. For good reason: it's specified explicitly. 

- So the whole drop-down-choose-FS makes sense only if there **isn't** a layout specified; the layout overrides the need to pick a FS. A distro that specifies a layout will not set any choices anyway, and it's not needed.
- **or** the layout can have a special *filesystem* value, say "default", that picks up .. the default! This can be implemented without the erase-choice code, even, so it would be testable with existing layouts and configurations. In the presence of a choosable FS on erase, that changes every item in the layout that is told to use the default FS.
- **or** there could be multiple layouts, one for each chosen FS on erase. How to specify this in the config is unclear; possibly a *partitionLayout[defaultfs]* kind of key would work, but then we need to store multiple layouts and pick the right one on erase.

**Comment by arcolinuxz** [comment]
Can this be incorporated in the next Calamares @adriaandegroot  or does it need further testing?
Always ready to be a betatester.

**Comment by LordTermor** [comment]
> Can this be incorporated in the next Calamares @adriaandegroot or does it need further testing?
> Always ready to be a betatester.

Don't have time to finish it properly unfortunately

**Comment by adriaandegroot** [comment]
This is now in branch *pr-1667* in the main repo, where I intend to resolve the issues that are still open in comments here.

**Comment by LordTermor** [comment]
> This is now in branch _pr-1667_ in the main repo, where I intend to resolve the issues that are still open in comments here.

Thank you a lot for taking it in your hands. PR can be closed I think?

**Comment by adriaandegroot** [comment]
> PR can be closed I think?

Let's leave it open until the branch is merged to *calamares*, otherwise I'll forget there's things pending (like the comments in this PR).

File: src/modules/partition/core/Config.cpp (hunk 5 above)
Comment (adriaandegroot) [reply src/modules/partition/core/Config.cpp:213]: I've decided to canonicalize (e.g. use the names that KPMCore uses internally).

**Comment by adriaandegroot** [comment]
(All the code-style and minor technical nits have been addressed in *pr-1667*, leaving the three "open conversation topics". I'm not going to get to them today, there's the mysterious-partition-flags issue to chase.

**Comment by adriaandegroot** [comment]
Topics **existing configuration** and **documentation** have been dealt with in the branch.

**Comment by adriaandegroot** [comment]
I have chosen to extend the partition-layout code -- from @gportay -- such that partitions may be specified as *unknown* and that that means "use the default / user selection", e.g. defers the **actual** choice of FS until later.
